src/llfs/ioring_log_device2.test.cpp

```diff
@@ -0,0 +1,594 @@
+
+      this->observed_trim = slot_range.lower_bound;
+      this->observed_flush = slot_range.upper_bound;
+
+      while (!this->maybe_flushed_slots.empty() &&
+             !llfs::slot_less_than(this->observed_flush,
+                                   this->maybe_flushed_slots.front().upper_bound)) {
+        this->observed_slot_flush = this->maybe_flushed_slots.front().upper_bound;
+        this->maybe_flushed_slots.pop_front();
+      }
+    }
+
+    /** \brief Attempts to reserve at least n_bytes for appending in the log, by trimming old slots
+     * if necessary.
+     *
+     * If this function returns true, all is fine; otherwise, it indicates a fatal error that should
+     * signal the end of the append phase of this scenario.
+     */
+    bool trim_to_reserve(usize n_bytes)
+    {
+      u64 observed_space = this->log_writer->space();
+      while (observed_space < n_bytes && !this->appended_slots.empty()) {
+        const llfs::SlotRange trimmed_slot = this->appended_slots.front();
+
+        this->appended_slots.pop_front();
+        this->maybe_trimmed_slots.push_back(trimmed_slot);
+
+        Status trim_status = this->log_device->trim(trimmed_slot.upper_bound);
+        if (!trim_status.ok()) {
+          break;
+        }
+
+        Status await_status = this->log_writer->await(llfs::BytesAvailable{
+            .size = trimmed_slot.size() + observed_space,
+        });
+        if (!await_status.ok()) {
+          break;
+        }
+
+        llfs::SlotRange slot_range = this->log_device->slot_range(llfs::LogReadMode::kDurable);
+        while (!this->maybe_trimmed_slots.empty() &&
+               llfs::slot_less_than(this->maybe_trimmed_slots.front().lower_bound,
+                                    slot_range.lower_bound)) {
+          LLFS_CHECK_SLOT_LE(this->maybe_trimmed_slots.front().upper_bound, slot_range.lower_bound);
+          observed_space += this->maybe_trimmed_slots.front().size();
+          this->maybe_trimmed_slots.pop_front();
+        }
+      }
+
+      // Update the observed trim/flush pointers.
+      //
+      this->update_trim_flush_pos();
+
+      // Return true iff we succeeded in trimming enough space.
+      //
+      return (observed_space >= n_bytes);
+    }
+
+    /** \brief Verify that the recovered durable slot range doesn't violate our assumptions:
+     *
+     * - observed trim pos should never move backwards
+     * - slot range should always be at most the total size of the log
+     * - slot range should not have negative size (i.e., lower_bound <= upper_bound)
+     * - the observed flush pos should either be the same as the last one observed, or it should be
+     *   not less than the highest observed slot upper bound for all appends.
+     */
+    void verify_recovered_slot_range(const llfs::SlotRange& slot_range)
+    {
+      ASSERT_TRUE(!llfs::slot_less_than(slot_range.upper_bound, slot_range.lower_bound))
```

Comment:
`ASSERT_TRUE` isn't like `BATT_ASSERT` or `assert` in that it is disabled on non-Debug builds.  It's part of the googletest API; it will return if the condition fails (so it must be used only in a context where the return type is `void`), and the currently running test will fail, even if the `ASSERT_TRUE` is called on a different thread.

Reply:
ahhhh makes sense. I definitely just read `ASSERT_TRUE` as `BATT_ASSERT`.